crypto/fipsmodule/hmac/hmac.c

```diff
@@ -118,62 +126,75 @@ struct hmac_methods_st {
+MD_TRAMPOLINES_EXPLICIT_PRECOMPUTED(SHA384, SHA512_CTX, SHA512_CBLOCK)
+MD_TRAMPOLINES_EXPLICIT_PRECOMPUTED(SHA512, SHA512_CTX, SHA512_CBLOCK)
+MD_TRAMPOLINES_EXPLICIT_PRECOMPUTED(SHA512_224, SHA512_CTX, SHA512_CBLOCK)
+MD_TRAMPOLINES_EXPLICIT_PRECOMPUTED(SHA512_256, SHA512_CTX, SHA512_CBLOCK)
+MD_TRAMPOLINES_EXPLICIT(SHA3_224, KECCAK1600_CTX, SHA3_224_CBLOCK)
+MD_TRAMPOLINES_EXPLICIT(SHA3_256, KECCAK1600_CTX, SHA3_256_CBLOCK)
+MD_TRAMPOLINES_EXPLICIT(SHA3_384, KECCAK1600_CTX, SHA3_384_CBLOCK)
+MD_TRAMPOLINES_EXPLICIT(SHA3_512, KECCAK1600_CTX, SHA3_512_CBLOCK)
 
 struct hmac_method_array_st {
   HmacMethods methods[HMAC_METHOD_MAX];
 };
 
+// This macro does not set any values for precomputed keys or state |h|, and
+// as such is suitable for use with Keccak/SHA3.
 #define DEFINE_IN_PLACE_METHODS(EVP_MD, HASH_NAME)  {                        \
     out->methods[idx].evp_md = EVP_MD;                                       \
-    out->methods[idx].chaining_length = HASH_NAME##_CHAINING_LENGTH;         \
     out->methods[idx].init = AWS_LC_TRAMPOLINE_##HASH_NAME##_Init;           \
     out->methods[idx].update = AWS_LC_TRAMPOLINE_##HASH_NAME##_Update;       \
     out->methods[idx].finalize = AWS_LC_TRAMPOLINE_##HASH_NAME##_Final;      \
-    out->methods[idx].init_from_state =                                      \
-        AWS_LC_TRAMPOLINE_##HASH_NAME##_Init_from_state;                     \
-    out->methods[idx].get_state = AWS_LC_TRAMPOLINE_##HASH_NAME##_get_state; \
+    out->methods[idx].chaining_length = 0UL;                                 \
+    out->methods[idx].init_from_state = NULL;                                \
+    out->methods[idx].get_state = NULL;                                      \
     idx++;                                                                   \
     assert(idx <= HMAC_METHOD_MAX);                                          \
   }
 
+// Use |idx-1| because DEFINE_IN_PLACE_METHODS has already incremented it.
+#define DEFINE_IN_PLACE_METHODS_PRECOMPUTED(EVP_MD, HASH_NAME)  {            \
+    DEFINE_IN_PLACE_METHODS(EVP_MD, HASH_NAME);                              \
+    out->methods[idx-1].chaining_length = HASH_NAME##_CHAINING_LENGTH;       \
+    out->methods[idx-1].init_from_state =                                    \
+        AWS_LC_TRAMPOLINE_##HASH_NAME##_Init_from_state;                     \
+    out->methods[idx-1].get_state =                                          \
+        AWS_LC_TRAMPOLINE_##HASH_NAME##_get_state;                           \
+  }
+
 DEFINE_LOCAL_DATA(struct hmac_method_array_st, AWSLC_hmac_in_place_methods) {
   OPENSSL_memset((void*) out->methods, 0, sizeof(out->methods));
   int idx = 0;
   // Since we search these linearly it helps (just a bit) to put the most common ones first.
   // This isn't based on hard metrics and will not make a significant different on performance.
   // FIXME: all hashes but SHA256 have been disabled to check first SHA256
```

Comment:
Do you know what this FIXME is referencing?

Reply:
I do not. It was added [with the pre-computed keys change](https://github.com/aws/aws-lc/commit/98ccf4a316401112943bed604562102ad52efac6). Do you know @fabrice102?